Avoid generating machine code repeatly for same basic block

qwe661234 opened the pull request:
Originally, every execution path had its own machine code. Therefore, even if some basic blocks had been traced in other execution paths, we still generated a new copy of machine code. To use the code cache space more efficiently, we modified the recorded table to be global, allowing every execution path to share the machine code. This modification ensures that every basic block has only one copy of machine code, thus saving code cache space.

For example, the code cache stored 1,926,471 (1.9 MB) bytes of machine code for AES previously. After the modification, the code cache stored 182,730 (0.18 MB) bytes of machine code.

jserv commented on src/jit.c:
```diff
@@ -1564,26 +1567,31 @@ static void translate_chained_block(struct jit_state *state,
+                return state->offset_map[i].offset;
+            }
+        }
+        assert(NULL);
```
Can it be replaced with `__UNREACHABLE`?

jserv commented on src/jit.c:
```diff
@@ -56,7 +56,8 @@
 
 #define STACK_SIZE 512
-#define MAX_INSNS 1024
+#define MAX_JUMPS 1024
```
Can you show empirical adjustments for both `MAX_JUMPS` and `MAX_INSNS`. I wonder how they can be calculated.

qwe661234 replied:
The `MAX_JUMPS` means the number of host branch instructions within a chained block collection need to be resolved.

qwe661234 replied:
I modified `MAX_INSNS` into `MAX_BLOCKS` because it refers to the total number of basic blocks within our code cache. Considering that our block cache size is 4096 and some blocks may be replaced, I set `MAX_BLOCKS` to 8192.

jserv commented on src/jit.c:
```diff
@@ -1531,29 +1532,30 @@ static void resolve_jumps(struct jit_state *state)
     }
 }
 
+set_t set;
```
Can it be declared as a local variable?

jserv commented on src/jit.c:
```diff
@@ -1349,7 +1350,7 @@ void parse_branch_history_table(struct jit_state *state, rv_insn_t *ir)
             max_idx = i;
     }
-    if (bt->PC[max_idx]) {
+    if (bt->PC[max_idx] && bt->times[max_idx] >= 256) {
```
Why 256 here?

qwe661234 replied:
As mentioned in the first commit of this PR, I set 256 as the threshold because we recorded 16 indirect jumps, and
the threshold for launching T1C is 4096. 256 == 4096 / 16. I think the indirect jump with the highest frequency should exceed this threshold.

jserv replied:
Avoid hardcoded value. Use certain identifier instead.

jserv reviewed (changes requested):
Drop the trailing dot in the subject of git commit messages.
See https://cbea.ms/git-commit/

qwe661234 committed:
Avoid generating machine code repeatly for same basic block

Originally, every execution path had its own machine code. Therefore,
even if some basic blocks had been traced in other execution paths, we
still generated a new copy of machine code. To use the code cache space
more efficiently, we modified the recorded table to be global, allowing
every execution path to share the machine code. This modification
ensures that every basic block has only one copy of machine code, thus
saving code cache space.

For example, the code cache stored 1,926,471 (1.9 MB) bytes of machine
code for AES previously. After the modification, the code cache stored
182,730 (0.18 MB) bytes of machine code.

jserv commented on src/jit.c:
```diff
@@ -1564,26 +1564,31 @@ static void translate_chained_block(struct jit_state *state,
                 max_idx = i;
         }
-        if (bt->PC[max_idx] && !set_has(set, bt->PC[max_idx])) {
+        if (bt->PC[max_idx] && bt->times[max_idx] >= 256 &&
```
Ditto. Avoid hardcoded value.

qwe661234 committed:
Avoid generating machine code for infrequently used indirect jumps

We set 256 as the threshold because we recorded 16 indirect jumps, and
the threshold for launching T1C is 4096. This modification also saves
code cache space by avoiding storing the machine code with less usage.
